_drafts/ember-orbit.md

```diff
@@ -0,0 +1,477 @@
+import Orbit from '@orbit/data';
+import fetch from 'fetch';
+
+export default {
+  create(injections = {}) {
+    // Use `fetch` implementation from `ember-fetch`
+    Orbit.fetch = fetch;
+
+    injections.name = 'remote';
+    injections.host = 'http://localhost:3000';
+
+    return new JSONAPISource(injections);
+  }
+};
+```
+
+Note that we need to override `Orbit.fetch`; otherwise we'll get an error.
+
+QUESTION: WHY DOES BUILT-IN FETCH PRODUCE THE FOLLOWING ERROR? TypeError: Failed to execute 'fetch' on 'Window': Illegal invocation
+
+We need to add `ember-fetch` to the project:
+
+```sh
+# ember install ember-fetch
+```
+
+Our `remote.js` adds the remote service, but it isn't automatically wired up to anything. We need to create a "strategy", a preconfigured Orbit object that configures multiple sources to talk to one another. Let's create a strategy that says that every time we update our in-memory store, we should update the remote store too:
+
+`app/data-strategies/remote-update.js`:
+
+```js
+import { RequestStrategy } from '@orbit/coordinator';
+
+export default {
+  create() {
+    return new RequestStrategy({
+      name: 'remote-update',
+
+      source: 'store',
+      on: 'beforeUpdate',
+
+      target: 'remote',
+      action: 'push',
+
+      blocking: true,
+    });
+  }
+};
+```
+
+This says that before every update to the store, we should push the corresponding changes to the remote store. “Blocking” means that the remote push must succeed before the local change is allowed to proceed; we’ll learn more about this option later.
+
+For strategies to run, we need to enable Orbit’s coordinator. It's created for us when our app starts, but not activated automatically; we'll see why later. Let's activate it in our application route:
+
+`app/routes/application.js`:
+
+```js
+import Route from '@ember/routing/route';
+import { inject as service } from '@ember/service';
+
+export default Route.extend({
+  dataCoordinator: service(),
+
+  beforeModel() {
+    const coordinator = this.dataCoordinator;
+    return coordinator.activate();
+  },
+});
+```
+
+Now if we run our app and check the network tab of our browser, every time we add a message we'll see an outgoing POST request to our API, and a 201 Created response:
+
+SCREENSHOT
+
+## EventLoggingStrategy: Console Output
+
+To see what’s happening in more detail, let's add event logging so we can see it in the browser console.
+
+`app/data-strategies/event-logging.js`:
+
+```js
+import { EventLoggingStrategy } from '@orbit/coordinator';
+
+export default {
+  create() {
+    return new EventLoggingStrategy();
+  }
+};
+```
+
+Now when you reload your app you'll see a few log entries immediately:
+
+- store beforeQuery
+- store query
+
+When you add a message, you'll see a number of more entries:
+
+- store beforeUpdate
+- remote beforePush
+- remote transform
+- remote push
+- store transform
+- store update
+
+DETAILS ON WHAT EACH OF THESE MEANS SPECIFICALLY
```

Comment:
?

Reply:
It is helpful to understand the different [data flows](http://orbitjs.com/v0.15/guide/data-flows.html) supported by Orbit. Different interfaces participate in the "request" vs. "sync" flows.

The `transform` event is emitted to reflect any transformation of state in a source. The `push` and `update` events are only emitted when the corresponding requests complete. Note that the `transform` event will also be emitted when these requests complete successfully.